### PR TITLE
En 3782/system vm and smart contracts

### PR DESCRIPTION
**Pull request by sasurobert**
Created a simple system smart contract executioner VM.
Created a container and a factory to create all the needed system smart contracts and store them by their name.
All the system smart contracts implement the same interface.
Created an environmental interface for smart contracts to interact with. They will write into a local cache, and in the end the VM will create a VMOutput of all the changed data and apply it to the trie - the same method used for Arwen and Hera. In case of an error the smart contract has to only signal the error, no need to do a complex revert as nothing is saved to the final trie.
Created interface for peer changes, a class is needed to implement it and change data accordingly.

Decisions to be made:
1. Should metachain process system smart contracts like register ? Or it is the job of the shard. If we put it to meta, meta will need interceptors, processor, pool for transactions. My suggestion is to make the shards process it - condition peer state must be global.

2. Interaction from the register smart contract to peer state to nodesCoordinator. Suggestion: shard processes registration, block is sent to meta, when metablock which includes the registration gets finalized, the peer state change is applied automatically in every shard -> nodesCoordinator can update its list according to the new peer states.

The PR is mostly a draft, no tests yet. Waiting for initial feedback in order not to rewrite too much.

Thanks.

### CODE REVIEW

File: vm/systemSmartContracts/register.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// Execute calls one of the functions from the register smart contract and runs the code according to the input
+func (r *registerSC) Execute(args *vm.ExecuteArguments) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	if args.Value.Cmp(r.stakeValue) != 0 {
+		return vmcommon.UserError
+	}
+
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/register.go]: I would have expected that the registration SC would have done the checking of the required balance of the sender. should have transferred toward the registration address the stake value and modify some internal values (that will then be sent toward VMOutput for permanent storage) that will track down the registration event. This vars might have been a map containing tuples like {address, metablock nonce, stake_value} or a variant like this.

File: vm/factory/systemSCFactory.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	"errors"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	"github.com/ElrondNetwork/elrond-go/vm/systemSmartContracts"
+	"math/big"
+)
+
+var initialStakeValue = "500000000000000000000000"
+
+type systemSCFactory struct {
+	systemEI vm.SystemEI
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCFactory.go]: an environment interface is what a smart contract interacts with. That is the general way of naming these things.

File: vm/interface.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	AddMultiple(keys [][]byte, scs []SystemSmartContract) error
+	Replace(key []byte, val SystemSmartContract) error
+	Remove(key []byte)
+	Len() int
+	Keys() [][]byte
+	IsInterfaceNil() bool
+}
+
+// SystemEI defines the environment interface system smart contract can use
+type SystemEI interface {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: systemEI is the the interface which can be used in a protocolSC.

File: vm/interface.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	Sender   []byte
+	Value    *big.Int
+	Function string
+	Args     []*big.Int
+}
+
+// SystemSmartContract interface defines the function a system smart contract should have
+type SystemSmartContract interface {
+	Execute(args *ExecuteArguments) vmcommon.ReturnCode
+	ValueOf(key interface{}) interface{}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: Ok. It can be deleted. I was thinking about a way the protocol questions a system smart contract about the value of a certain element.

File: vm/interface.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	Len() int
+	Keys() [][]byte
+	IsInterfaceNil() bool
+}
+
+// SystemEI defines the environment interface system smart contract can use
+type SystemEI interface {
+	Transfer(destination []byte, sender []byte, value *big.Int, input []byte) error
+	BlockHash(number int64) []byte
+	GetVMInput() vmcommon.VMInput
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: This is a minimal number of functions the SC need to interact with. In case of ethereum is even bigger, for Arwen is big as well. Should the SC know about a number of interfaces ? It is easier for it to interact with only one. This is the basic library, for specific functionalities we will have a new EI.

File: vm/interface.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	GetBalance(addr []byte) *big.Int
+	SetStorage(addr []byte, key []byte, value []byte)
+	GetStorage(addr []byte, key []byte) []byte
+	GetSCAddress() []byte
+	SelfDestruct(addr []byte, beneficiary []byte)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: This is a standard interface used in many blockchain projects. We might use it when migrating a protocol SC to another VM.

File: vm/process/systemVM.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	inputArgs := &vm.ExecuteArguments{
+		Sender:   input.CallerAddr,
+		Value:    input.CallValue,
+		Function: input.Function,
+		Args:     input.Arguments,
+	}
+	returnCode := contract.Execute(inputArgs)
+
+	if returnCode == vmcommon.Ok {
+		//TODO: save to the trie all the changes which are for other than account trie
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/process/systemVM.go]: Agreed for normal contracts, but what about peer state changed ? Should I put the processing of peer changes into SmartContractProcessor ? Here in system VM at least we know the SC is authorized to make changes in the peerState. IN the smartContractProcessor we have to treat SCs differently according to address, whether they are allowed to make changes in peer state or not.

File: vm/systemSmartContracts/common.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package systemSmartContracts
+
+import (
+	"errors"
+	"github.com/ElrondNetwork/elrond-go/vm"
+)
+
+func CheckIfNil(args *vm.ExecuteArguments) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/common.go]: It will be used by every protocolSC.

File: vm/interface.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+// SystemSCContainerFactory defines the functionality to create a system smart contract container
+type SystemSCContainerFactory interface {
+	Create() (SystemSCContainer, error)
+	IsInterfaceNil() bool
+}
+
+// SystemSCContainer defines a system smart contract holder data type with basic functionality
+type SystemSCContainer interface {
+	Get(key []byte) (SystemSmartContract, error)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: I just copied the way other containers are here.

File: vm/process/systemVM.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/ElrondNetwork/elrond-go/vm"
+	"github.com/ElrondNetwork/elrond-go/vm/factory"
+	"github.com/ElrondNetwork/elrond-go/vm/systemSmartContracts"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+type systemVM struct {
+	blockchainHook vmcommon.BlockchainHook
+	cryptoHook     vmcommon.CryptoHook
+	acnts          state.AccountsAdapter
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/process/systemVM.go]: Normal accounts state will be deleted, as it is not needed. But still not decided who does the final writing into PeerState - SmartContractProcessor vs SystemVM.

File: vm/process/systemVM.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	scFactory, err := factory.NewSystemSCFactory(vm.systemEI)
+	if err != nil {
+		return nil, err
+	}
+
+	vm.systemContracts, err = scFactory.Create()
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/process/systemVM.go]: It will be changed to be injected at constructor. Here I exemplified all the components, for the first reviews.

File: vm/process/systemVM.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	vm := &systemVM{
+		blockchainHook: hook,
+		cryptoHook:     crypto,
+	}
+	copy(vm.vmType, vmType)
+
+	var err error
+	vm.systemEI, err = systemSmartContracts.NewVMContext(vm.blockchainHook, vm.cryptoHook)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/process/systemVM.go]: It will be. It was the draft for the first review, to see if the architecture is well designed.

File: vm/process/systemVM.go [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+func (s *systemVM) RunSmartContractCall(input *vmcommon.ContractCallInput) (*vmcommon.VMOutput, error) {
+	s.systemEI.CleanCache()
+	s.systemEI.SetContractCallInput(input)
+
+	contract, err := s.systemContracts.Get(input.RecipientAddr)
+	if err != nil || contract.IsInterfaceNil() {
+		return nil, vm.ErrUnknownSystemSmartContract
+	}
+
+	inputArgs := &vm.ExecuteArguments{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/process/systemVM.go]: I have to change the vmCommon definition of VMInput and put sender and receiver address into the structure - as those are missing.

File: vm/systemSmartContracts/register.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/register.go]: Balance is checked by smartContract processor and the minus part is taken out of the sender address balance before executing a smart contract. Here the registration has to write into a buffer/adapter/cache which will update after correct execution the peer state.

File: vm/process/systemVM.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply vm/process/systemVM.go]: If we do it in this manner, we will end up injecting PeerState in all VM variants that we will support (because we might think to run the protocol SCs in other VMs) So yes, that's the protocol's job to interpret the results in a meaningful way. What happens if we need to write in a new trie or a new area? We need to inject that new area also in all VMs so we could run a piece of a SC. I'm sticking to the idea that VMs should only execute stuff, not caring about where to save/load the data.We might have the peer processor interpret the stored data through a VM (in this way we keep the decoupling and the flexibility of simply swapping the VMs): just like I did with ScDataGetter.

File: vm/systemSmartContracts/common.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply vm/systemSmartContracts/common.go]: Should add comment then.

File: vm/process/systemVM.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply vm/process/systemVM.go]: Definitely not the VM. As explained above, this can be done by the peer processor through a ScDataGetter.

File: vm/process/systemVM.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply vm/process/systemVM.go]: ok

File: vm/process/systemVM.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply vm/process/systemVM.go]: ok

File: vm/process/systemVM.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply vm/process/systemVM.go]: We might add those fields in VmCommon if we can get rid of this argument.

File: vm/factory/systemSCContainer.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Replace will add (or replace if it already exists) an object at a given key
+func (vmc *systemSCContainer) Replace(key []byte, sc vm.SystemSmartContract) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCContainer.go]: The container structure is copied from another container, which has the same naming.

File: vm/factory/systemSCFactory_test.go [hunk 15]
```diff
@@ -0,0 +1 @@
+package factory
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCFactory_test.go]: done.

File: vm/process/systemVM.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	if crypto != nil {
+		return nil, vm.ErrNilCryptoHook
+	}
+
+	vm := &systemVM{
+		blockchainHook: hook,
+		cryptoHook:     crypto,
+	}
+	copy(vm.vmType, vmType)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/process/systemVM.go]: done.

File: vm/process/systemVM.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/process/systemVM.go]: changed and deleted the executeArguments structure.

File: vm/systemSmartContracts/register.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+type registerSC struct {
+	eei        vm.SystemEI
+	stakeValue *big.Int
+}
+
+// NewRegisterSmartContract creates a register smart contract
+func NewRegisterSmartContract(stakeValue *big.Int, eei vm.SystemEI) (*registerSC, error) {
+	if stakeValue == nil {
+		return nil, errors.New("initial stake value is nil")
```
Comment (BeniaminDrasovean) [inline vm/systemSmartContracts/register.go]: Move error to errors.go.

File: vm/systemSmartContracts/register.go [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	if eei == nil || eei.IsInterfaceNil() {
+		return nil, errors.New("environment interface is nil")
```
Comment (BeniaminDrasovean) [inline vm/systemSmartContracts/register.go]: Use vm.ErrNilSystemEnvironmentInterface.

File: process/factory/factory.go [hunk 19]
```diff
@@ -19,8 +19,11 @@ const (
...
-// InternalTestingVM is a byte array identified for the smart contract address created for the testing VM
-var InternalTestingVM = []byte{0, 0}
+// SystemVirtualMachine is a byte array identifier for the smart contract address created for system VM
+var SystemVirtualMachine = []byte{0, 1}
 
 // IELEVirtualMachine is a byte array identifier for the smart contract address created for IELE VM
 var IELEVirtualMachine = []byte{1, 0}
+
+// InternalTestingVM is a byte array identified for the smart contract address created for the testing VM
+var InternalTestingVM = []byte{255, 255}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/factory/factory.go]: done

File: vm/factory/systemSCContainer.go [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+	if !ok {
+		return nil, process.ErrWrongTypeInContainer
+	}
+
+	return sc, nil
+}
+
+// Add will add an object at a given key. Returns
+// an error if the element already exists
+func (vmc *systemSCContainer) Add(key []byte, sc vm.SystemSmartContract) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCContainer.go]: good catch. done.

File: vm/factory/systemSCContainer.go [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+	if !ok {
+		return process.ErrContainerKeyAlreadyExists
+	}
+
+	return nil
+}
+
+// Replace will add (or replace if it already exists) an object at a given key
+func (vmc *systemSCContainer) Replace(key []byte, sc vm.SystemSmartContract) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCContainer.go]: done.

File: vm/factory/systemSCFactory.go [hunk 22]
```diff
@@ -0,0 +1,53 @@
+package factory
+
+import (
+	"errors"
+	"github.com/ElrondNetwork/elrond-go/vm"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCFactory.go]: done.

File: vm/factory/systemSCFactory.go [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+	return &systemSCFactory{systemEI: systemEI}, nil
+}
+
+// Create instantiates all the system smart contracts and returns a container
+func (scf *systemSCFactory) Create() (vm.SystemSCContainer, error) {
+	scContainer := NewSystemSCContainer()
+
+	initValue, ok := big.NewInt(0).SetString(initialStakeValue, 10)
+	if !ok {
+		return nil, errors.New("bad config value for initial stake")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCFactory.go]: done

File: vm/interface.go [hunk 24]
```diff
@@ -0,0 +1,54 @@
+package vm
+
+import (
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"math/big"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: done

File: vm/interface.go [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+	Len() int
+	Keys() [][]byte
+	IsInterfaceNil() bool
+}
+
+// SystemEI defines the environment interface system smart contract can use
+type SystemEI interface {
+	Transfer(destination []byte, sender []byte, value *big.Int, input []byte) error
+	GetBalance(addr []byte) *big.Int
+	SetStorage(addr []byte, key []byte, value []byte)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: this is the typical API used by any smart contract on ethereum or other blockchain platforms. dApp do not care how the storage is kept, he just reads and write key value data, from his own memory / storage space.

File: vm/systemSmartContracts/common.go [hunk 26]
```diff
@@ -0,0 +1,31 @@
+package systemSmartContracts
+
+import (
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+func CheckIfNil(args *vmcommon.ContractCallInput) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/common.go]: done

File: vm/systemSmartContracts/eei.go [hunk 27]
```diff
@@ -0,0 +1,195 @@
+package systemSmartContracts
+
+import (
+	"fmt"
+	"github.com/ElrondNetwork/elrond-go/vm"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: done.

File: vm/systemSmartContracts/eei.go [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+	"math/big"
+
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+type SCFunctionHandler func(value *big.Int, args []*big.Int)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: deleted as not used.

File: vm/systemSmartContracts/eei.go [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+	strAdr := string(addr)
+	if _, ok := host.storageUpdate[strAdr]; ok {
+		if value, ok := host.storageUpdate[strAdr][string(key)]; ok {
+			return value
+		}
+	}
+
+	data, err := host.blockChainHook.GetStorageData(addr, key)
+	if err != nil {
+		fmt.Printf("GetStorage returned with error %s \n", err.Error())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: this is mostly an error supposed to help the development of the contract, the protocol does not uses it, so I deleted it.

File: vm/systemSmartContracts/eei.go [hunk 30]
```diff
@@ -0,0 +1,195 @@
...
+func (host *vmContext) GetBalance(addr []byte) *big.Int {
+	strAdr := string(addr)
+	if outAcc, ok := host.outputAccounts[strAdr]; ok {
+		actualBalance := big.NewInt(0).Add(outAcc.Balance, outAcc.BalanceDelta)
+		return actualBalance
+	}
+
+	balance, err := host.blockChainHook.GetBalance(addr)
+	if err != nil {
+		fmt.Printf("GetBalance returned with error %s \n", err.Error())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: same as above. deleted

File: vm/systemSmartContracts/eei.go [hunk 31]
```diff
@@ -0,0 +1,195 @@
...
+		BalanceDelta: big.NewInt(0),
+		Address:      addr}
+
+	return balance
+}
+
+// Transfer handles any necessary value transfer required and takes
+// the necessary steps to create accounts
+func (host *vmContext) Transfer(destination []byte, sender []byte, value *big.Int, input []byte,
+) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: rearranged.

File: vm/systemSmartContracts/eei.go [hunk 32]
```diff
@@ -0,0 +1,195 @@
...
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+type SCFunctionHandler func(value *big.Int, args []*big.Int)
+
+type vmContext struct {
+	blockChainHook vmcommon.BlockchainHook
+	cryptoHook     vmcommon.CryptoHook
+
+	storageUpdate  map[string](map[string][]byte)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: done.

File: vm/systemSmartContracts/register.go (hunk 17 above)
Comment (sasurobert) [reply vm/systemSmartContracts/register.go]: done

File: vm/systemSmartContracts/register.go (hunk 18 above)
Comment (sasurobert) [reply vm/systemSmartContracts/register.go]: done.

File: vm/factory/systemSCContainer.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply vm/factory/systemSCContainer.go]: Don't forget to add the test for this case.

File: vm/factory/systemSCContainer.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply vm/factory/systemSCContainer.go]: Add a test for the nil/empty key situation

File: vm/interface.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply vm/interface.go]: ok.

File: vm/systemSmartContracts/register.go [hunk 33]
```diff
@@ -1,13 +1,27 @@
 package systemSmartContracts
 
 import (
-	"errors"
+	"bytes"
+	"encoding/json"
+	"github.com/ElrondNetwork/elrond-go/core/logger"
```
Comment (BeniaminDrasovean) [inline vm/systemSmartContracts/register.go]: Sort imports

File: vm/process/systemVM_test.go [hunk 34]
```diff
@@ -0,0 +1,90 @@
...
+	assert.Equal(t, vm.ErrUnknownSystemSmartContract, err)
+}
+
+func TestSystemVM_RunSmartContractCall(t *testing.T) {
+	t.Parallel()
+
+	sc := &mock.SystemSCStub{ExecuteCalled: func(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+		return vmcommon.Ok
+	}}
+	scAddress := []byte("tralala")
```
Comment (iulianpascalau) [inline vm/process/systemVM_test.go]: :D

File: vm/systemSmartContracts/eei.go [hunk 35]
```diff
@@ -0,0 +1,201 @@
...
+	if !ok {
+		destAcc = &vmcommon.OutputAccount{
+			Address:      destination,
+			BalanceDelta: big.NewInt(0),
+			Balance:      big.NewInt(0),
+		}
+		host.outputAccounts[string(destAcc.Address)] = destAcc
+	}
+
+	senderAcc.BalanceDelta = big.NewInt(0).Sub(senderAcc.BalanceDelta, value)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/eei.go]: You can call 
```GO
senderAcc.BalanceDelta.Sub(senderAcc.BalanceDelta, value)
```
Valid also for the next line

File: vm/systemSmartContracts/eei.go [hunk 36]
```diff
@@ -0,0 +1,201 @@
...
+	for addr := range host.selfDestruct {
+		vmOutput.DeletedAccounts = append(vmOutput.DeletedAccounts, []byte(addr))
+	}
+
+	// save to the output finally
+	for _, outAcc := range outAccs {
+		vmOutput.OutputAccounts = append(vmOutput.OutputAccounts, outAcc)
+	}
+
+	vmOutput.GasRemaining = big.NewInt(0)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/eei.go]: Is this supposed to be this way? We might have a fixed gas for these VM operations. It is not fair to refund 0 when calling a protocol SC func as you might want to put a large gas limit so you are assured that the operation is done.

File: vm/systemSmartContracts/staking.go [hunk 37]
```diff
@@ -0,0 +1,226 @@
...
+// NewStakingSmartContract creates a staking smart contract
+func NewStakingSmartContract(stakeValue *big.Int, eei vm.SystemEI) (*stakingSC, error) {
+	if stakeValue == nil {
+		return nil, vm.ErrNilInitialStakeValue
+	}
+	if eei == nil || eei.IsInterfaceNil() {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+
+	reg := &stakingSC{stakeValue: big.NewInt(0).Set(stakeValue), eei: eei}
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go]: make these multi-line

File: vm/systemSmartContracts/staking.go [hunk 38]
```diff
@@ -0,0 +1,226 @@
...
+		StartNonce:    0,
+		Staked:        false,
+		BlsPubKey:     nil,
+		UnStakedNonce: 0,
+		StakeValue:    big.NewInt(0),
+	}
+	data := r.eei.GetStorage(args.CallerAddr)
+
+	if data != nil {
+		err := json.Unmarshal(data, registrationData)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go]: might want to inject marshalizer interface instead of using json

File: vm/systemSmartContracts/staking.go [hunk 39]
```diff
@@ -0,0 +1,226 @@
...
+			log.Error("unmarshal error on finalize unstake function" + err.Error())
+			return vmcommon.UserError
+		}
+
+		if registrationData.UnStakedNonce == 0 {
+			log.Error("validator did not unstaked yet")
+			return vmcommon.UserError
+		}
+
+		r.eei.SetStorage(arg.Bytes(), nil)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go]: Might also add a check here to not be able to call finalize un stake before n blocks passed.

File: vm/systemSmartContracts/staking.go [hunk 40]
```diff
@@ -0,0 +1,226 @@
...
+			log.Error("transfer error on finalizeUnStake function " + err.Error())
+			return vmcommon.UserError
+		}
+	}
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) slash(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	ownerAddress := r.eei.GetStorage([]byte(ownerKey))
+	if !bytes.Equal(ownerAddress, args.CallerAddr) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go]: I do not think slash can work like this. Who is the owner that can slash one node?

File: vm/systemSmartContracts/staking_test.go [hunk 41]
```diff
@@ -0,0 +1 @@
+package systemSmartContracts
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking_test.go]: missing tests. Might be left to a next PR after we decide this is the structure that we want.

File: vm/process/systemVM.go [hunk 42]
```diff
@@ -0,0 +1,94 @@
...
+
+	return sVm, nil
+}
+
+// RunSmartContractCreate creates and saves a new smart contract to the trie
+func (s *systemVM) RunSmartContractCreate(input *vmcommon.ContractCreateInput) (*vmcommon.VMOutput, error) {
+	if input == nil {
+		return nil, vm.ErrInputArgsIsNil
+	}
+	if input.CallerAddr == nil {
```
Comment (iulianpascalau) [inline vm/process/systemVM.go]: Might improve code coverage on this func

File: vm/factory/factory.go [hunk 43]
```diff
@@ -0,0 +1,5 @@
+package factory
+
+// TODO: work on creating these addresses
+
+var RegisterSCAddress = []byte("12345678901234567890123456789012")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/factory.go]: done

File: vm/factory/systemSCContainer.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCContainer.go]: done

File: vm/factory/systemSCContainer.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCContainer.go]: done

File: vm/systemSmartContracts/register.go (hunk 33 above)
Comment (sasurobert) [reply vm/systemSmartContracts/register.go]: done

File: vm/systemSmartContracts/eei.go (hunk 35 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: linter will say that output is not used :|.

File: vm/systemSmartContracts/eei.go (hunk 36 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: currently there are no gas usage on systemVM, when it is added, this line will be changed.

File: vm/systemSmartContracts/staking.go (hunk 39 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go]: These things are saved in a local cache, so if you return with userError, non of the cached values are taken into consideration, TxValue is sent back to the sender, gas is consumed. That code is in SmartContract processor.

File: vm/systemSmartContracts/staking.go (hunk 38 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go]: It will be much harder to move the code to another VM.

File: vm/systemSmartContracts/staking.go (hunk 40 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go]: The owner of the smart contract is predefined, so any metachain node can do it. But the block is only signed when there is no wrong slashing. Otherwise people might just call from the own addresses the slashing function, and you do not know at the VM level if it was a good call or not. Deciding whether it is a good call or not it is part of peer processor, block creation.

File: vm/systemSmartContracts/staking_test.go (hunk 41 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking_test.go]: wanted to leave for the next PR as code is a lot.

Comment (sasurobert) [reply vm/systemSmartContracts/staking_test.go]: and also might give the writing of unit tests to a junior.

File: vm/systemSmartContracts/eei.go [hunk 44]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+func (host *vmContext) SetContractCallInput(input *vmcommon.ContractCallInput) {
+	host.callFunction = input.Function
+	host.scAddress = input.RecipientAddr
+	host.vmInput = input.VMInput
+}
+
+// SelfDestruct destroy the current smart contract, setting the beneficiary for the liberated storage fees
+func (host *vmContext) SelfDestruct(addr []byte, beneficiary []byte) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: we might, when moving it to another VM and the system is already running.

[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: you have to kill unused items

File: vm/systemSmartContracts/eei.go [hunk 45]
```diff
@@ -0,0 +1,218 @@
+package systemSmartContracts
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: Elrond Environment Interface ... it is long :dagger:

File: vm/interface.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: it is an API, not GO style coding. should conform with other API's.

File: vm/interface.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/interface.go]: check again as it is :dagger:

File: vm/factory/systemSCFactory.go [hunk 46]
```diff
@@ -0,0 +1,50 @@
+package factory
+
+import (
+	"errors"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	"github.com/ElrondNetwork/elrond-go/vm/systemSmartContracts"
+	"math/big"
+)
+
+var initialStakeValue = "500000000000000000000000"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/factory/systemSCFactory.go]: agreed. Will be added with the new PR.

File: vm/systemSmartContracts/eei.go (hunk 35 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: changed.

File: vm/systemSmartContracts/staking.go (hunk 37 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go]: done.

File: vm/process/systemVM.go (hunk 42 above)
Comment (sasurobert) [reply vm/process/systemVM.go]: done
